File: py/D10.py
def parseInput():
    file = open("../inputs/D10.input", "r")
    jolts = []
    contents = file.read()
    for line in contents.splitlines():
        jolts.append(int(line))
    return jolts
    
def p1():
    jolts = parseInput()
    jolts.append(0)
    jolts.append(max(jolts) + 3)
    jolts.sort()

    ones = 0
    threes = 0

    for i in range(len(jolts) - 1):
        diff = jolts[i + 1] - jolts[i]
        if diff == 1: ones += 1
        elif diff == 3: threes += 1

    return ones * threes

def p2():
    jolts = parseInput()
    jolts.append(0)
    jolts.append(max(jolts) + 3)
    jolts.sort()
    dp = {}
    dp[len(jolts)-1] = 0
    dp[len(jolts)-2] = 1

    for i in reversed(range(len(jolts) - 2)):
        dp[i] = 0

        if jolts[i + 1] - jolts[i] <= 3:
            dp[i] += dp[i + 1]
        
        if jolts[i + 2] - jolts[i] <= 3:
            dp[i] += dp[i + 2]
        
        if i + 3 < len(jolts) and jolts[i + 3] - jolts[i] <= 3:
            dp[i] += dp[i + 3]

    return dp[0]

print(p1())
print(p2())

File: ts/D10.ts
import { readFileSync } from 'fs';
import { timeAndPrint } from './time-and-print';

function parseInput(): number[] {
    const numbers = [];
    const content = readFileSync('../inputs/D10.input', 'utf8');
    for (let line of content.split("\r\n")) {
        numbers.push(+line);
    }
    return numbers;
}

function p1() {
    const jolts = parseInput();
    jolts.push(0);
    jolts.push(Math.max(...jolts) + 3);
    jolts.sort((a, b) => a - b);
    let ones = 0, threes = 0;

    for (let i = 0; i + 1 < jolts.length; i++) {
        const diff = jolts[i + 1] - jolts[i];
        switch (diff) {
            case 1: ones++; break;
            case 3: threes++; break;
            default: continue;
        }
    }

    return ones * threes;
}

function p2() {
    const jolts = parseInput();
    jolts.push(0);
    jolts.push(Math.max(...jolts) + 3);
    jolts.sort((a, b) => a - b);

    const dp = new Array<number>(jolts.length);
    dp[jolts.length - 1] = 0;
    dp[jolts.length - 2] = 1;

    for (let i = jolts.length - 3; i >= 0; i--) {
        let count = 0;
        if (jolts[i + 1] - jolts[i] <= 3)
            count += dp[i + 1];
        if (jolts[i + 2] - jolts[i] <= 3)
            count += dp[i + 2];
        if (i + 3 < jolts.length && jolts[i + 3] - jolts[i] <= 3)
            count += dp[i + 3];
        dp[i] = count;
    }

    return dp[0];
}

timeAndPrint(p1);
timeAndPrint(p2);